src/repr/src/row/encoding.rs

```diff
@@ -95,213 +114,41 @@ impl From<&ScalarType> for ColumnFormat {
 }
 
 /// A helper for adapting mz's [Datum] to persist's columnar [Data].
-#[derive(Debug)]
-pub enum DatumEncoder<'a> {
-    Bool(&'a mut <bool as Data>::Mut),
-    BoolOpt(&'a mut <Option<bool> as Data>::Mut),
-    UInt16(&'a mut <u16 as Data>::Mut),
-    UInt16Opt(&'a mut <Option<u16> as Data>::Mut),
-    UInt32(&'a mut <u32 as Data>::Mut),
-    UInt32Opt(&'a mut <Option<u32> as Data>::Mut),
-    UInt64(&'a mut <u64 as Data>::Mut),
-    UInt64Opt(&'a mut <Option<u64> as Data>::Mut),
-    Int16(&'a mut <i16 as Data>::Mut),
-    Int16Opt(&'a mut <Option<i16> as Data>::Mut),
-    Int32(&'a mut <i32 as Data>::Mut),
-    Int32Opt(&'a mut <Option<i32> as Data>::Mut),
-    Int64(&'a mut <i64 as Data>::Mut),
-    Int64Opt(&'a mut <Option<i64> as Data>::Mut),
-    Float32(&'a mut <f32 as Data>::Mut),
-    Float32Opt(&'a mut <Option<f32> as Data>::Mut),
-    Float64(&'a mut <f64 as Data>::Mut),
-    Float64Opt(&'a mut <Option<f64> as Data>::Mut),
-    String(&'a mut <String as Data>::Mut),
-    StringOpt(&'a mut <Option<String> as Data>::Mut),
-    Bytes(&'a mut <Vec<u8> as Data>::Mut),
-    BytesOpt(&'a mut <Option<Vec<u8>> as Data>::Mut),
-    Todo(&'a mut <Vec<u8> as Data>::Mut),
+pub struct DatumEncoder<'a>(Box<dyn FnMut(Datum) + 'a>);
+
+impl std::fmt::Debug for DatumEncoder<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DatumEncoder").finish_non_exhaustive()
+    }
 }
 
 impl<'a> DatumEncoder<'a> {
+    fn new<T: Data>(
+        part: &mut ColumnsMut<'a>,
+        name: &ColumnName,
+        from_datum: for<'b> fn(Datum<'b>) -> T::Ref<'b>,
```

Comment:
if you make this generic (so `where for<'b> F: Fn(Datum<'b>) -> T::Ref<'b>`) I think you will recoup from a second indirection, as the closure will be moved into `Box<dyn FnMut>` as opposed to the function pointer